Sources/SwiftFusion/Inference/FactorBoilerplate.swift.gyb

```diff
@@ -0,0 +1,154 @@
+  return ', '.join([template(i) for i in range(arity)])
+
+def tuple_component(tuple, i):
+  return ''.join([tuple] + ['.tail'] * i + ['.head'])
+}%
+
+// Artifact of Swift weakness.
+/// Do not use this. Use `Factor${arity}` instead.
+public protocol Factor${arity}_ {
+  % for i in range(arity):
+  /// The ${i}-th variable type.
+  associatedtype V${i}
+  % end
+
+  /// Returns the error at the given point.
+  ///
+  /// This is typically interpreted as negative log-likelihood.
+  func error(${repeated('_: V%d')}) -> Double
+}
+
+/// A factor in a factor graph.
+public protocol Factor${arity}: Factor, Factor${arity}_
+  where Variables == Tuple${arity}<${repeated('V%d')}> {}
+
+extension Factor${arity} {
+  % for i in range(arity):
+  /// The variable vertex for this factor's ${i}-th variable.
+  public var input${i}ID: TypedID<V${i}, Int> { return ${tuple_component('edges', i)} }
+  % end
+
+  // Forwarding implementation.
+  public func error(at x: Variables) -> Double {
+    return error(${repeated(lambda i: tuple_component('x', i))})
+  }
+}
+
+// Artifact of Swift weakness.
+/// Do not use this. Use `VectorFactor${arity}` instead.
+public protocol VectorFactor${arity}_ {
+  % for i in range(arity):
+  /// The ${i}-th variable type.
+  associatedtype V${i}
+  % end
+
+  typealias Variables = Tuple${arity}<${repeated('V%d')}>
+
+  associatedtype ErrorVector: EuclideanVectorN
+  associatedtype LinearizableComponent: LinearizableFactor
+
+  /// Returns the error vector at the given point.
+  func errorVector(${repeated('_: V%d')}) -> ErrorVector
+
+  /// Returns the linearizable component of `self` at the given point, and returns the
+  /// `Differentiable` subset of the given variables.
+  func linearizableComponent(${repeated('_: V%d')})
+    -> (LinearizableComponent, LinearizableComponent.Variables)
+}
+
+/// A factor whose `error` is a function of a vector-valued `errorVector` function.
+public protocol VectorFactor${arity}: VectorFactor, VectorFactor${arity}_
+  where Variables == Tuple${arity}<${repeated('V%d')}> {}
+
+extension VectorFactor${arity} {
+  % for i in range(arity):
+  /// The variable vertex for this factor's ${i}-th variable.
+  public var input${i}ID: TypedID<V${i}, Int> { return ${tuple_component('edges', i)} }
+  % end
+
+  // Implements the error as the squared norm of the error vector.
+  public func error(at x: Variables) -> Double {
+    return errorVector(at: x).squaredNorm
+  }
+
+  // Forwarding implementation.
+  public func errorVector(at x: Variables) -> ErrorVector {
+    return errorVector(${repeated(lambda i: tuple_component('x', i))})
+  }
+
+  // Forwarding implementation.
+  public func linearizableComponent(at x: Variables)
+    -> (LinearizableComponent, LinearizableComponent.Variables)
+  {
+    return linearizableComponent(${repeated(lambda i: tuple_component('x', i))})
+  }
+}
+
+// Artifact of Swift weakness.
+/// Do not use this. Use `LinearizableFactor${arity}` instead.
```

Comment:
This line is identical with the next line, a bit confusing?

Reply:
Not sure what you mean. Maybe you haven't noticed the underscore at the end of `public protocol LinearizableFactor${arity}_`?